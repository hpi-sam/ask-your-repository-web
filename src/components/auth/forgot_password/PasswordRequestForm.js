// @flow
import React, { useState } from 'react';
import { useDispatch } from 'react-redux';
import { push } from 'connected-react-router';
import { flashErrorMessage, flashSuccessMessage } from 'redux-flash';
import { Link } from 'react-router-dom';
import Form from 'react-validation/build/form';
import Input from 'react-validation/build/input';
import Button from 'react-validation/build/button';
import ValidationErrors from '../../utility/form/ValidationErrors';
import type { Errors } from '../../../models/Errors';
import AuthService from '../../../services/AuthService';

const PasswordRequestForm = () => {
  const dispatch = useDispatch();
  const [email, setEmail] = useState('');
  const [errors, setErrors] = useState<Errors>({ missingInput: false });

  const resetErrors = () => {
    setErrors({
      missingInput: false,
    });
  };

  const handleSubmit = async (e: SyntheticEvent<HTMLFormElement>) => {
    e.preventDefault();
    resetErrors();

    if (email) {
      try {
        await AuthService.requestResetLink(email);
        dispatch(flashSuccessMessage('Successfully requested a password request. Please check your emails for the reset link.'));
        dispatch(push('/login'));
      } catch (error) {
        const message = error.response
          ? error.response.data.error
          : 'Could not establish a connection to the server.';

        dispatch(flashErrorMessage(message));
      }
    } else {
      setErrors({ missingInput: true });
    }
  };

  const printError = () => {
    const error = Object.keys(errors).find(key => errors[key]);

    return error
      ? <ValidationErrors error={error} />
      : null;
  };

  return (
    <Form onSubmit={handleSubmit} className="Form Form__centered">
      <div className="Form__title">Reset Password</div>
      {printError()}
      <div className="Form__input">
        <label className="Form__input__label">
          Email or Username:
          <Input
            type="text"
            name="email"
            value={email}
            onChange={e => setEmail(e.target.value)}
          />
        </label>
      </div>
      <div className="Form__buttons">
        <Button className="Form__buttons__item Form__buttons__item__blue">
          Submit
        </Button>
        <Link to="/login" className="Form__buttons__item Form__buttons__item__gray">Cancel</Link>
      </div>
    </Form>
  );
};

export default PasswordRequestForm;
