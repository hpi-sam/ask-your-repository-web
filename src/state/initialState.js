// @flow

// This file is only for state related testing

import type { AppState } from './AppState';
import { initialState as presentation } from './presentation/presentation.reducer';
import { initialState as teamSidebar } from './team_sidebar/teamSidebar.reducer';
import { initialState as presentationMode } from './presentation_mode/presentationMode.reducer';
import { initialState as activeTeam } from './active_team/activeTeam.reducer';
import { initialState as auth } from './auth/auth.reducer';

const initialState: AppState = {
  flash: { messages: [] },
  presentation,
  activeTeam,
  teamSidebar,
  presentationMode,
  auth,
};

export default initialState;
