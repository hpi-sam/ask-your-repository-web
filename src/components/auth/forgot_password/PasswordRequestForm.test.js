// @flow
import React from 'react';
import toJson from 'enzyme-to-json';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { mount } from 'enzyme';
import Form from 'react-validation/build/form';
import ValidationErrors from '../../utility/form/ValidationErrors';
import PasswordRequestForm from './PasswordRequestForm';
import initialState from '../../../state/initialState';
import AuthService from '../../../services/AuthService';

const mockStore = configureMockStore();

jest.mock('../../../services/AuthService');


describe('<PasswordRequestForm />', () => {
  let wrapper;
  let store;

  beforeAll(() => {
    AuthService.requestResetLink.mockImplementation(() => Promise.resolve([]));
  });

  beforeEach(() => {
    store = mockStore(initialState);
    wrapper = mount((
      <Provider store={store}>
        <MemoryRouter>
          <PasswordRequestForm />
        </MemoryRouter>
      </Provider>
    )).find('PasswordRequestForm');
  });

  afterEach(() => {
    AuthService.requestResetLink.mockClear();
  });

  it('renders correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('shows a validation Error when submitting without providing an email', () => {
    wrapper.find(Form).simulate('submit');
    wrapper = wrapper.update();
    expect(wrapper.find(ValidationErrors).text()).toEqual('Please fill in all fields.');
  });

  it('calls the auth service with the provided email', () => {
    const testEmail = 'test@example.com';
    wrapper.find('input').simulate('change', { target: { name: 'email', value: testEmail } });
    wrapper.find(Form).simulate('submit');
    expect(AuthService.requestResetLink).toHaveBeenCalledWith(testEmail);
  });
});
