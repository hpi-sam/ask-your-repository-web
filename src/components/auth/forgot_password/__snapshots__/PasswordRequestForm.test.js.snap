// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PasswordRequestForm /> renders correctly 1`] = `
<PasswordRequestForm>
  <Form(t)
    className="Form Form__centered"
    onSubmit={[Function]}
  >
    <t
      className="Form Form__centered"
      getValues={[Function]}
      hideError={[Function]}
      onSubmit={[Function]}
      showError={[Function]}
      validate={[Function]}
      validateAll={[Function]}
    >
      <form
        className="Form Form__centered"
        onSubmit={[Function]}
      >
        <div
          className="Form__title"
        >
          Reset Password
        </div>
        <div
          className="Form__input"
        >
          <label
            className="Form__input__label"
          >
            Email or Username:
            <Control(s)
              name="email"
              onChange={[Function]}
              type="text"
              validations={Array []}
              value=""
            >
              <s
                name="email"
                onBlur={[Function]}
                onChange={[Function]}
                type="text"
                value=""
              >
                <div>
                  <input
                    name="email"
                    onBlur={[Function]}
                    onChange={[Function]}
                    type="text"
                    value=""
                  />
                </div>
              </s>
            </Control(s)>
          </label>
        </div>
        <div
          className="Form__buttons"
        >
          <Button(s)
            className="Form__buttons__item Form__buttons__item__blue"
          >
            <s
              className="Form__buttons__item Form__buttons__item__blue"
              hasErrors={false}
            >
              <button
                className="Form__buttons__item Form__buttons__item__blue"
                disabled={false}
              >
                Submit
              </button>
            </s>
          </Button(s)>
          <Link
            className="Form__buttons__item Form__buttons__item__gray"
            to="/login"
          >
            <a
              className="Form__buttons__item Form__buttons__item__gray"
              href="/login"
              onClick={[Function]}
            >
              Cancel
            </a>
          </Link>
        </div>
      </form>
    </t>
  </Form(t)>
</PasswordRequestForm>
`;
